txpool/txpool.go

```diff
@@ -471,60 +468,164 @@ func (t *TxPool) ResetWithHeader(h *types.Header) {
 	t.ProcessEvent(evnt)
 }
 
-// ProcessEvent processes the blockchain event and resets the txpool accordingly
+// ProcessEvent processes the blockchain event and updates the txpool accordingly
 func (t *TxPool) ProcessEvent(evnt *blockchain.Event) {
-	addTxns := map[types.Hash]*types.Transaction{}
-	for _, evnt := range evnt.OldChain {
-		// reinject these transactions on the pool
-		block, ok := t.store.GetBlockByHash(evnt.Hash, true)
-		if !ok {
-			t.logger.Error("block not found on txn add", "hash", block.Hash())
-		} else {
-			for _, txn := range block.Transactions {
-				addTxns[txn.Hash] = txn
-			}
-		}
-	}
-
-	delTxns := map[types.Hash]*types.Transaction{}
+	txnsInBlocks := map[types.Hash]*types.Transaction{}
 	for _, evnt := range evnt.NewChain {
-		// remove these transactions from the pool
+		// Grab the block that has just been written to state
 		block, ok := t.store.GetBlockByHash(evnt.Hash, true)
 		if !ok {
 			t.logger.Error("block not found on txn del", "hash", block.Hash())
 		} else {
+			// Compile transactions that should be accounted for in the TxPool
 			for _, txn := range block.Transactions {
-				delete(addTxns, txn.Hash)
-				delTxns[txn.Hash] = txn
+				txnsInBlocks[txn.Hash] = txn
 			}
 		}
 	}
 
-	// try to include again the transactions in the pendingQueue list
-	for _, txn := range addTxns {
-		if err := t.addImpl(OriginReorg, txn); err != nil {
-			t.logger.Error("failed to add txn", "err", err)
-		}
+	// txDropCleanup is a helper method for updating the gauge size,
+	// as well as removing leftover remote txns
+	txnDropCleanup := func(txn *types.Transaction) {
+		t.gauge.decrease(slotsRequired(txn))
+		t.remoteTxns.Delete(txn)
 	}
 
-	// remove the mined transactions from the pendingQueue list
-	for _, txn := range delTxns {
+	// dropTxnCallback is a helper function for
+	// removing a transaction from the pending queue
+	// If the transaction is present in the pending queue,
+	// drop it, drop it from the remoteTxns as well, and decrease the slot
+	// it takes up
+	dropTxnCallback := func(txn *types.Transaction) {
 		if ok := t.pendingQueue.Delete(txn); ok {
-			t.gauge.decrease(slotsRequired(txn))
-			t.remoteTxns.Delete(txn)
+			t.logger.Debug(
+				fmt.Sprintf("Dropping txn [%s] from the promoted queue",
+					txn.Hash.String(),
+				),
+			)
+
+			txnDropCleanup(txn)
 		}
 	}
-	//update the metric
+
+	// Remove the txns from the block that were just committed to state
+	// from any queues in the TxPool
+	for _, txn := range txnsInBlocks {
+		// Initially drop any txns that are leftover in the TxPool queues,
+		// but are submitted to the chain state
+		dropTxnCallback(txn)
+
+		// Attempt to update the next account nonce in the TxPool
+		// with the one in state if it's greater
+		wrapper := t.lockAccountQueue(txn.From, true)
+
+		// Grab the latest nonce from state
+		stateRoot := t.store.Header().StateRoot
+		stateNonce := t.store.GetNonce(stateRoot, txn.From)
```

Comment:
How long does it take time to call them? If the block has multiple transactions for one account, you would query for state nonce of the account multiple times. You might want to create nonce map first.

Reply:
It's not a super-costly operation, but it does add additional overhead.

I've refactored the way we do this state nonce fetching to be more optimal 👍 

01f412d